EZP-31413: Migrating a project to v3 (#953)
ezsystems/developer-documentation: docs/updating/upgrading_to_v3.md

> # Upgrading to v3

Reviewer: ```suggestion
# Upgrading eZ Platform to v3
```

To upgrade your v2 project to v3, you need to follow the [standard update procedure](../updating/updating_ez_platform.md),
but you will also need to make changes to your code.

## 1\. Automatically refactor your code

> To simplify the process of adapting your code to Symfony 5, you can use [rector](https://github.com/rectorphp/rector)

Reviewer: ```suggestion
To simplify the process of adapting your code to Symfony 5, you can use [Rector, a reconstructor tool](https://github.com/rectorphp/rector)
```

> to automatically refactor your Symfony and PHPunit code.

Reviewer: ```suggestion
that will automatically refactor your Symfony and PHPunit code.
```

## 2\. Adapt project structure

!!! tip

    For details on all changes related to the switch to Symfony 5,

>     see [Symfony upgrade guide for 4.0](https://github.com/symfony/symfony/blob/master/UPGRADE-4.0.md)

Reviewer: ```suggestion
    see [Symfony upgrade guide for 4.0](https://github.com/symfony/symfony/blob/5.0/UPGRADE-4.0.md)
```

>     and [for 5.0](https://github.com/symfony/symfony/blob/master/UPGRADE-5.0.md)

Reviewer: ```suggestion
    and [for 5.0](https://github.com/symfony/symfony/blob/5.0/UPGRADE-5.0.md)
```

The latest Symfony versions changed the organization of your project into folders and bundles.
When updating to eZ Platform v3 you need to move your files and modify file paths and namespace references.

![Project structure changes in v3](img/folder_structure_v3.png "Project folder structure changes between v2 and v3")

#### Configuration

Configuration files have been moved from `app/Resources/config` to `config`.
Package-specific configuration is placed in `config/packages` (e.g. `config/packages/ezplatform_admin_ui.yaml`).

> This folder also contains `config/packages/ezplatform.yaml`, which contains all settings coming in from the Kernel.

Reviewer: ```suggestion
This folder also contains `config/packages/ezplatform.yaml`, which contains all settings coming in from Kernel.
```

#### PHP code and bundle organization

Since Symfony 4 your code is no longer organized in bundles.
`AppBundle` has been removed from eZ Platform.

> You need to move all your PHP code, such as controllers or event listeners, in the `src` folder.

Reviewer: ```suggestion
You need to move all your PHP code, such as controllers or event listeners, to the `src` folder.
```

Use the `App` namespace for your custom code instead.

#### View templates

Templates are no longer stored in `app/Resources/views`.

> You need to move all your templates in the `templates` folder in your project's root.

Reviewer: ```suggestion
You need to move all your templates to the `templates` folder in your project's root.
```

#### Translations

Translation files have been moved out of `app/Resources/translations` into `translations` in your project's root.

#### `web` and assets

Content of the `web` folder is now placed in `public`.
Content of `app/Resources/assets` has been moved to `assets`.

!!! note

    You also need to update paths that refer to the old location,

>     for example in [`webpack.config.js`](../guide/bundles.md#configuration-from-a-bundle)

Reviewer: ```suggestion
    for example in [`webpack.config.js`](../guide/bundles.md#configuration-from-a-bundle).
```

|---|---|
|EzSystems\RepositoryForms\Data\Content\FieldData|EzSystems\EzPlatformContentForms\Data\Content\FieldData|
|EzSystems\RepositoryForms\Data\FieldDefinitionData|EzSystems\EzPlatformAdminUi\Form\Data\FieldDefinitionData|
|EzSystems\RepositoryForms\FieldType\FieldDefinitionFormMapperInterface|EzSystems\EzPlatformAdminUi\FieldType\FieldDefinitionFormMapperInterface|
|EzSystems\RepositoryForms\FieldType\FieldValueFormMapperInterface|EzSystems\EzPlatformContentForms\FieldType\FieldValueFormMapperInterface|

#### Extending Field Type templates

If you extended templates for `ezobjectrelationlist_field`, `ezimageasset_field`, or `ezobjectrelation_field` Fields
using `{% extends "@EzPublishCore/content_fields.html.twig" %}`,
you now need to extend `EzSystemsPlatformHttpCache` instead, if you wish to make use of cache:
`{% extends "@EzSystemsPlatformHttpCache/content_fields.html.twig" %}`.

### Signal Slots and Events

If you used Signal Slots to listen for events in you custom code,
you need to rewrite them using Symfony Events and Listeners instead.

The application now triggers two Events per operation: one before and one after the relevant thing happens

> (see for example [Bookmark events](https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/Core/Event/BookmarkService.php)).

Reviewer: ```suggestion
(see for example [Bookmark events](https://github.com/ezsystems/ezplatform-kernel/blob/v1.0.0/eZ/Publish/Core/Event/BookmarkService.php)).
```

> To use them, create [Event Listeners](https://symfony.com/doc/4.3/event_dispatcher.html) in your code,

Reviewer: ```suggestion
To use them, create [Event Listeners](https://symfony.com/doc/5.0/event_dispatcher.html) in your code,
```

to use [custom actions](../extending/extending_workflow.md#adding-custom-actions).

To migrate your content which had been using Flex Workflow to the new Quick Review workflow,
run the following command:

`php bin/console ezplatform:migrate:flex-workflow`

### Universal Discovery Widget

If you extended the Universal Discovery Widget
(e.g. added your own tabs or triggered opening the UDW for your own customizations),
you need to rewrite this extension using the [new YAML configuration](../extending/extending_udw.md).

### Back Office extensibility

If you added custom tab groups in the Back Office,
you now need to [make use of the `TabsComponent`](../guide/extending/extending_tabs.md#adding-a-new-tab-group).

### HTTP cache

HTTP cache bundle now uses FOS Cache Bundle v2.
If your code makes use of HTTP cache bundle, see [the list of changes and deprecations](../releases/ez_platform_v3.0_deprecations.md#http-cache-bundle).

### REST

If your code extends the REST API, you need to modify namespaces.
The `eZ\Publish\Core\REST` and `eZ\Publish\Core\REST\Common\` namespaces have been replaced by `EzSystems\EzPlatformRest`.

> This is due to the fact that REST code has been moved from kernel to a new `ezpublish-rest` package.

Reviewer: ```suggestion
This is due to the fact that REST code has been moved from Kernel to a new `ezpublish-rest` package.
```

**Use:**

``` php
services:
    Acme\App\Installer\MyCustomInstaller:
        parent: EzSystems\PlatformInstallerBundle\Installer\DbBasedInstaller
```

**instead of:**

``` php
services:
    Acme\App\Installer\MyCustomInstaller:
        parent: ezplatform.installer.db_based_installer
```

### Other changes

#### User checker

Add the user checker to firewall by adding the following line to `config/packages/security.yaml`:

``` yaml hl_lines="5"
security:
    firewalls:
        ezpublish_front:
            # ...
            user_checker: eZ\Publish\Core\MVC\Symfony\Security\UserChecker
            # ...
```

#### Command

If your custom commands use `Symfony\Bundle\FrameworkBundle\Command\ContainerAwareCommand`
you need to rewrite them to use `Symfony\Component\Console\Command\Command` instead,
because `ContainerAwareCommand` is deprecated.

#### Permission-related methods

Some deprecated [permission-related methods](../releases/ez_platform_v3.0_deprecations.md#permission-related-methods) have been removed.
If your code uses them, you need to rewrite it to use the permission resolver.

#### Container parameters

> A numer of Symfony Dependency Injection Container parameters [have been dropped](https://github.com/ezsystems/ezpublish-kernel/blob/master/doc/bc/8.0/dropped-container-parameters.md).

Reviewer: ```suggestion
A numer of Symfony Dependency Injection Container parameters [have been dropped](https://github.com/ezsystems/ezplatform-kernel/blob/v1.0.0/doc/bc/1.0/dropped-container-parameters.md).
```

To check if your code uses them, search for all occurrences of `ezpublish\..*\.class` (regular expression pattern)

> and [decorate Symfony services](https://symfony.com/doc/4.4/service_container/service_decoration.html) instead.

Reviewer: ```suggestion
and [decorate Symfony services](https://symfony.com/doc/5.0/service_container/service_decoration.html) instead.
```

and [decorate Symfony services](https://symfony.com/doc/4.4/service_container/service_decoration.html) instead.

#### QueryTypes

If your code relies on automatically registering QueryTypes through the naming convention `<Bundle>\QueryType\*QueryType`,
you need to register your QueryTypes as services and tag them with `ezpublish.query`, or enable their automatic configuration (`autoconfigure: true`).

#### Symfony namespaces

The following Symfony namespaces have changed. You need to update your code if it uses any of them:

|Use|Instead of|
|---|---|
|Symfony\Contracts\Translation\TranslatorInterface|Symfony\Component\Translation\TranslatorInterface|
|Symfony\Contracts\EventDispatcher\Event|Symfony\Component\EventDispatcher\Event|

#### Deprecations

For a full list of changes, see [Deprecations and backwards compatibility breaks](../releases/ez_platform_v3.0_deprecations.md)

### Update the database

Apply the following database update script:

`mysql -u <username> -p <password> <database_name> < vendor/ezsystems/ezpublish-kernel/blob/master/data/update/mysql/dbupdate-7.5.latest-to-8.0.0.sql`

or for PostgreSQL:

> `psql <database_name> < vendor/ezsystems/ezpublish-kernel/blob/master/data/update/postgres/dbupdate-7.5.latest-to-8.0.0.sql`

Reviewer 2: These links won't work, the scripts are now here: https://github.com/ezsystems/ezplatform/tree/master/upgrade/db

!!! enterprise

> 

Reviewer 2: And for Enterprise they are here: https://github.com/ezsystems/ezplatform-ee/tree/master/upgrade/db

so maybe it's not worth keeping them here?

    ```sql
    DROP TABLE IF EXISTS ezsite;
    
    CREATE SEQUENCE ezsite_seq;
    
    CREATE TABLE ezsite (
      id int NOT NULL DEFAULT NEXTVAL ('ezsite_seq'),
      name varchar(255) NOT NULL DEFAULT '',
      created int NOT NULL,
      PRIMARY KEY (id)
    ) ;
    
    --
    --
    
    DROP TABLE IF EXISTS ezsite_public_access;
    
    CREATE TABLE ezsite_public_access (
      public_access_identifier varchar(255) NOT NULL,
      site_id int NOT NULL,
      site_access_group varchar(255) NOT NULL DEFAULT '',
      status int NOT NULL,
      config text NOT NULL,
      site_matcher_host varchar(255) DEFAULT NULL,
      PRIMARY KEY (public_access_identifier),
      CONSTRAINT fk_ezsite_public_access_site_id FOREIGN KEY (site_id) REFERENCES ezsite (id)
    ) ;
    
    CREATE INDEX ezsite_public_access_site_id ON ezsite_public_access (site_id);
    ```

### Update Apache/Nginx configuration

Make sure that you Apache/Nginx configuration is up to date with Symfony 5.
Refer to [the provided `vhost.template`](https://github.com/ezsystems/ezplatform/blob/master/doc/apache2/vhost.template)
for an example.

> ## 5\. Update the database

Reviewer 2: Now I see that this section: https://github.com/ezsystems/developer-documentation/pull/953/files?short_path=919ebf3#diff-877068fcb53101273f10ca25b5750738R439 should be probably removed?
